pylabrobot/centrifuge/agilent_centrifuge.py

```diff
@@ -0,0 +1,787 @@
+    await self.send(b"\xaa\x01\x19\x28\x42")
+    await self.send(b"\xaa\x01\x0e\x0f")
+
+    resp = "89"
+    while resp == "89":
+      await self.com()
+      resp = await self.status_check()
+      resp = f"{resp[0]:02x}"
+
+    await self.send(b"\xaa\x01\x0e\x0f")
+    await self.send(b"\xaa\x01\x0e\x0f")
+
+    await self.send(b"\xaa\x01\x17\x02\x1a")
+    await self.send(b"\xaa\x01\x0e\x0f")
+    await self.send(b"\xaa\x01\xe6\xc8\x00\xb0\x04\x96\x00\x0f\x00\x4b\x00\xa0\x0f\x05\x00\x07")
+    await self.send(b"\xaa\x01\x17\x04\x1c")
+    await self.send(b"\xaa\x01\x17\x01\x19")
+
+    await self.send(b"\xaa\x01\x0b\x0c")
+    await self.send(b"\xaa\x01\x0e\x0f")
+    await self.send(b"\xaa\x01\xe6\xc8\x00\xb0\x04\x96\x00\x0f\x00\x4b\x00\xa0\x0f\x05\x00\x07")
+    # send 17 byte - bytes index 4:7 = homing position + 8000
+    st = self.homing_position + 8000
+    byte_string = st.to_bytes(4, byteorder="little")
+    await self.send(b"\xaa\x01\xd4\x97" + byte_string + b"\xc3\xf5\x28\x00\xd7\x1a\x00\x00\x49")
+    await self.send(b"\xaa\x01\x0e\x0f")
+    await self.send(b"\xaa\x01\x0e\x0f")
+
+    resp = "08"
+    while resp != "09":
+      resp = await self.send(b"\xaa\x01\x0e\x0f")
+      await self.status_check()
+      resp = f"{resp[0]:02x}"
+    await self.send(b"\xaa\x01\x0e\x0f")
+    await self.send(b"\xaa\x01\x0e\x0f")
+
+    await self.send(b"\xaa\x01\x17\x02\x1a")
+
+    await self.com()
+    await self.lock_door()
+
+  async def status_check(self):
+    """Check the status of the centrifuge."""
+    resp = await self.send(b"\xaa\x01\x0e\x0f")
+    # s = []
+    # for byte in resp:
+    #   s.append(f"{byte:02x}")
+    # await self.com()
+    # self.position = int.from_bytes(bytes.fromhex("".join(s[1:4])), byteorder="little")
+    # self.homing_position = int.from_bytes(bytes.fromhex("".join(s[9:12])), byteorder="little")
+    # stat = f"{resp[0]:02x}"
+    # print(f"status: {stat} position: {self.position}  homing position: {self.homing_position}")
+    return resp
+
+  async def open_door(self):
+    await self.send(b"\xaa\x02\x26\x00\x07\x2f")
+    await self.com()
+
+  async def close_door(self):
+    await self.send(b"\xaa\x02\x26\x00\x05\x2d")
+    await self.com()
+
+  async def lock_door(self):
+    await self.send(b"\xaa\x02\x26\x00\x01\x29")
+    await self.com()
+
+  async def unlock_door(self):
+    await self.send(b"\xaa\x02\x26\x00\x05\x2d")
+    await self.com()
+
+  async def lock_bucket(self):
+    await self.send(b"\xaa\x02\x26\x00\x07\x2f")
+    await self.com()
+
+  async def unlock_bucket(self):
+    await self.send(b"\xaa\x02\x26\x00\x06\x2e")
+    await self.com()
+
+  async def go_to_bucket1(self):
+    new_position = (self.position + 4000).to_bytes(4, byteorder="little")
+    byte_string = b"\xaa\x01\xd4\x97" + new_position + b"\xc3\xf5\x28\x00\xd7\x1a\x00\x00"
+    sum_byte = (sum(byte_string)-0xaa)&0xff
+    byte_string += sum_byte.to_bytes(1, byteorder="little")
+    tx_payloads = [
```

Comment:
omg, this is intense

Reply:
Yes Phi has been cooking (his first reverse engineering project)